### PR TITLE
Add methods that automatically split proto requests into batches and …

### PR DESCRIPTION
**Pull request by mjanuszkiewicz-tt**
…merge them back

### CODE REVIEW

File: src/main/java/com/traveltime/sdk/dto/requests/TimeFilterFastProtoRequest.java [hunk 1]
```diff
@@ -87,4 +90,35 @@
             .<TravelTimeError>mapLeft(cause -> new IOError(cause, IO_PROTO_ERROR + cause.getMessage()))
             .flatMap(this::parseResponse);
     }
+
+    static public List<TimeFilterFastProtoRequest> split(TimeFilterFastProtoRequest request, int batchSizeHint) {
+        /* Naively splitting requests may lead to situations where the last request is very small and inefficient.
+         * We adjust the batch sizes to never have a situation where a batch is smaller than (loadFactor * batchSizeHint).
+         */
+
+        float loadFactor = 0.1f;
```
Comment (mjanuszkiewicz-tt) [inline src/main/java/com/traveltime/sdk/dto/requests/TimeFilterFastProtoRequest.java:99]: Should this be user-configurable?

File: src/main/java/com/traveltime/sdk/TravelTimeSDK.java [hunk 2]
```diff
@@ -109,14 +113,56 @@
...
+    public Either<TravelTimeError, TimeFilterFastProtoResponse> sendProtoBatched(
+        TimeFilterFastProtoRequest request,
+        int batchSizeHint
+    ) {
+        return getFuture(sendProtoAsyncBatched(request, batchSizeHint));
+    }
+
+    public CompletableFuture<Either<TravelTimeError, TimeFilterFastProtoResponse>> sendProtoAsyncBatched(
+        TimeFilterFastProtoRequest request,
+        int batchSizeHint
```
Comment (mjanuszkiewicz-tt) [inline src/main/java/com/traveltime/sdk/TravelTimeSDK.java:140]: Should we have a default batch size of, say, 100k?

Comment (Donatas-L) [reply src/main/java/com/traveltime/sdk/TravelTimeSDK.java:140]: I think we should default to 100k. We have benchmarked with 100k in mind.

**Comment by Donatas-L** [comment]
@PKJonas Can you benchmark this changes?

**Comment by PKJonas** [comment]
> @PKJonas Can you benchmark this changes?
I'll give it a shot, keep you guys posted.

File: src/main/java/com/traveltime/sdk/dto/requests/proto/OneToMany.java [hunk 3]
```diff
@@ -8,6 +8,7 @@
 @Value
 @Builder
 @AllArgsConstructor
+@With
```
Comment (bartosz-witkowski) [inline src/main/java/com/traveltime/sdk/dto/requests/proto/OneToMany.java:11]: Does this do anything?

File: src/main/java/com/traveltime/sdk/TravelTimeSDK.java (hunk 2 above)
Comment (bartosz-witkowski) [reply src/main/java/com/traveltime/sdk/TravelTimeSDK.java:140]: We've benchmarked with 100k as the ceiling, the optimal default batch size might be something different.

File: src/main/java/com/traveltime/sdk/dto/requests/proto/OneToMany.java (hunk 3 above)
Comment (mjanuszkiewicz-tt) [reply src/main/java/com/traveltime/sdk/dto/requests/proto/OneToMany.java:11]: Generates `withField` methods that create a copy of the object with the `field` set to whatever you provide in the argument. In this case I need `withDestinationCoordinates`

File: src/main/java/com/traveltime/sdk/dto/requests/TimeFilterFastProtoRequest.java (hunk 1 above)
Comment (bartosz-witkowski) [reply src/main/java/com/traveltime/sdk/dto/requests/TimeFilterFastProtoRequest.java:100]: I think for now a sensible default is enough.

**Pull request merged**